Userland/Libraries/LibJIT/ELFBuild.cpp

```diff
@@ -0,0 +1,137 @@
+{
+    Checked<u64> final_image_size = sizeof(Elf64_Ehdr);
+    Vector<u64> section_offsets;
+    section_offsets.ensure_capacity(sections.size());
+
+    auto const sections_begin = final_image_size.value_unchecked();
+    final_image_size += sizeof(Elf64_Shdr) * sections.size();
+
+    for (auto const& section : sections) {
+        auto const offset = final_image_size.value();
+        section_offsets.unchecked_append(offset);
+        if (section.data.has_value()) {
+            final_image_size += section.data.value().size();
+        }
+    }
+
+    auto image = MUST(FixedArray<u8>::create(final_image_size.value()));
+
+    {
+        auto section_headers = Span<Elf64_Shdr> {
+            reinterpret_cast<Elf64_Shdr*>(image.span().offset_pointer(sections_begin)),
+            sections.size(),
+        };
+        for (size_t i = 0; i < sections.size(); ++i) {
+            section_headers[i] = sections[i].header;
+            section_headers[i].sh_offset = section_offsets[i];
+            if (sections[i].data.has_value()) {
+                auto const data = sections[i].data.value();
+                auto const data_in_elf = image.span().slice(section_offsets[i], data.size());
+                data.copy_to(data_in_elf);
+                section_headers[i].sh_size = data.size();
+            }
+        }
+    }
+
+    {
+        auto* const final_elf_hdr = reinterpret_cast<Elf64_Ehdr*>(image.data());
+        final_elf_hdr->e_ident[EI_MAG0] = 0x7f;
+        final_elf_hdr->e_ident[EI_MAG1] = 'E';
+        final_elf_hdr->e_ident[EI_MAG2] = 'L';
+        final_elf_hdr->e_ident[EI_MAG3] = 'F';
+        final_elf_hdr->e_ident[EI_CLASS] = ELFCLASS64;
+        // FIXME: This is platform-dependent. Any big-endian host will write the data in
+        // MSB format, so the EI_DATA field should be set to ELFDATA2MSB.
+        final_elf_hdr->e_ident[EI_DATA] = ELFDATA2LSB;
+        final_elf_hdr->e_ident[EI_VERSION] = EV_CURRENT;
+        // XXX: This is platform-dependent. The host must set the OSABI to the
+        // same it's being executed in. I'm not sure if this matters for GDB
+        // compatibility, though.
+        final_elf_hdr->e_ident[EI_OSABI] = ELFOSABI_SYSV;
+        final_elf_hdr->e_ident[EI_ABIVERSION] = 0;
+        auto padding = Bytes {
+            &final_elf_hdr->e_ident[EI_PAD],
+            EI_NIDENT - EI_PAD,
+        };
+        padding.fill(0);
+
+        // Set it to an "object" file, as GDB seems to request it. See
+        // https://sourceware.org/gdb/current/onlinedocs/gdb.html/Registering-Code.html#Registering-Code
+        // for more context.
+        final_elf_hdr->e_type = ET_REL;
+        // FIXME: This platform-dependent. This must be set to the host
+        // architecture.
+        final_elf_hdr->e_machine = EM_AMD64;
+        final_elf_hdr->e_version = EV_CURRENT;
+
+        // XXX: If we somehow end up needing to specify code sections in the
```

Comment:
Skeptical on these `XXX`-style  comments. hopefully they're not direct copy pasta from V8 :sweat_smile:. At the very least they should be re-written as FIXMEs.

Reply:
I put them there while prototyping thinking that at some point this ought to be moved to LibELF. Since I managed to separate the GDB part from the ELF building part, seems wise to move these build utilities to LibELF, although they don't support segments.

Reply:
Sure, I got that those should be FIXMEs, I don't discuss style. I just didn't take the "copying from V8" comment right 😄